simulator/src/main/java/com/github/benmanes/caffeine/cache/simulator/policy/irr/ClockProPlusPolicy.java

```diff
@@ -0,0 +1,663 @@
+  }
+
+  private void validateClockStructure() {
+    checkState(listHead != null);
+
+    if (handHot == null) {
+      checkState(sizeHot == 0);
+    } else {
+      checkState(handHot.isHot());
+      for (Node n = listHead.prev; n != handHot; n = n.prev) {
+        checkState(!n.isInTest());
+        checkState(n != handTest);
+      }
+    }
+
+    if (handCold == null) {
+      checkState(sizeResCold == 0);
+    } else {
+      checkState(handCold.isResidentCold());
+      for (Node n = listHead.prev; n != handCold; n = n.prev) {
+        checkState(!n.isResidentCold());
+      }
+    }
+
+    if (handTest == null) {
+      checkState(sizeInTest == 0);
+    } else {
+      checkState(handTest.isInTest());
+      for (Node n = listHead.prev; n != handTest; n = n.prev) {
+        checkState(n.isResident());
+        checkState(!n.isInTest());
+      }
+    }
+  }
+
+  private void validateStatus() {
+    checkState(listHead != null);
+
+    int sizeHot;
+    int sizeResCold;
+    int sizeNonResCold;
+    int sizeInTest;
+    int sizeRecentlyDemoted;
+    sizeHot = sizeInTest = sizeResCold = sizeNonResCold = sizeRecentlyDemoted = 0;
+
+    Node node = listHead;
+    do {
+      if (node == null) {
+        break;
+      }
+      checkState(node.isInClock());
+      if (node.isHot()) {
+        sizeHot++;
+      }
+      if (node.isResidentCold()) {
+        sizeResCold++;
+      }
+      if (!node.isResident()) {
+        sizeNonResCold++;
+      }
+      if (node.isInTest()) {
+        sizeInTest++;
+      }
+      if (node.demoted) {
+        sizeRecentlyDemoted++;
+      }
+      node = node.next;
+    } while (node != listHead);
+
+    checkState(sizeHot == this.sizeHot);
+    checkState(sizeNonResCold == this.sizeNonResCold);
+    checkState(sizeInTest == this.sizeInTest);
+    checkState(sizeResCold == this.sizeResCold);
+    checkState(sizeRecentlyDemoted == this.sizeDemoted);
+    checkState(sizeHot + sizeResCold == maxSize - sizeFree);
+    checkState(sizeResCold + sizeFree >= minResColdSize);
+    checkState(sizeResCold <= maxResColdSize);
+    checkState(sizeNonResCold <= maxNonResSize);
+  }
+
+  /** Prints out the internal state of the policy. */
+  private void printClock() {
+    System.out.println("** CLOCK-Pro list HEAD (small recency) **");
+    System.out.println(listHead.toString());
+    for (Node n = listHead.next; n != listHead; n = n.next) {
+      System.out.println(n.toString());
+    }
+    System.out.println("** CLOCK-Pro list TAIL (large recency) **");
+  }
+
+  // +----- Status ------+- Resident -+- In Test -+
+  // |               HOT |       TRUE |     FALSE |
+  // |          COLD_RES |       TRUE |     FALSE |
+  // |  COLD_RES_IN_TEST |       TRUE |      TRUE |
+  // |      COLD_NON_RES |      FALSE |      TRUE |
+  // |      OUT_OF_CLOCK |      FALSE |     FALSE |
+  // +-------------------+------------+-----------+
+  enum Status {
+    HOT, COLD_RES, COLD_RES_IN_TEST, COLD_NON_RES, OUT_OF_CLOCK,
+  }
```

Comment:
fyi, no changes expected.

A nice Java-specific feature is that enums can carry both state and methods. This table could be encoded into the value, e.g.

```java
enum Status {
  HOT(/* resident */ true, /* in test */ false),
  ...

  final boolean resident;
  final boolean inTest;

  Status(boolean resident, boolean inTest) {
    this.resident = resident;
    this.inTest = inTest;
  }

  public boolean isInTest() {
    return inTest;
  }
  ...
}
```

Reply:
Thank you for the comment! I didn't make change but that approach seems clearer and easier to use!